### PR TITLE
AV-210361 NodePortLocal support for AKO GatewayAPI

### PR DESCRIPTION
**Pull request by aaha97**

### CODE REVIEW

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (pkoshtavmware) [inline tests/gatewayapitests/npltests/gateway_npl_test.go:452]: cleanupGatewayForNPL is already inclusive of delete service

**Review by pkoshtavmware — COMMENTED** [review]
Overall unit test coverage is nice for NPL. 
Have added few comments.
Also please add details of end-to-end testing done

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:85]: no, we don't need pod to service here.

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:439]: we are not deleting the mapping here as the mapping is required for pod to httproute

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:452]: added code to remove duplicates

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:481]: added code to remove duplicates

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:530]: added code to remove duplicates

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/gateway_model_rel.go:439]: The comment `"Pod not found, deleting mappings"` is misleading

**Review by pkoshtavmware — COMMENTED** [review]
Overall PR looks good to me. Please add details of end-to-end testing done

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (akshayhavile) [inline tests/gatewayapitests/npltests/gateway_npl_test.go:133]: Ports are 60000+ now for NPL.

File: ako-gateway-api/k8s/gateway_controller.go
Comment (aaha97) [reply ako-gateway-api/k8s/gateway_controller.go]: removed call to IncrementQueueCounter

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (aaha97) [reply tests/gatewayapitests/npltests/gateway_npl_test.go]: updated port value.

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

File: ako-gateway-api/lib/lib.go
Comment (aaha97) [reply ako-gateway-api/lib/lib.go:38]: fixed

**Comment by aaha97** [comment]
build ako

File: ako-gateway-api/k8s/gateway_controller.go
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/gateway_controller.go:336]: During the delete event , will pod get deleted or only annotations will get deleted ?

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (pkoshtavmware) [inline ako-gateway-api/nodes/gateway_model_rel.go:463]: One basic question for my understanding why services from oldserviceslist are being added to servicesList ?

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (pkoshtavmware) [inline tests/gatewayapitests/npltests/gateway_npl_test.go:127]: This might also need to be changed to 6000x

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (pkoshtavmware) [inline tests/gatewayapitests/npltests/gateway_npl_test.go:141]: Port value change mighe be required here as well

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:439]: deletion of mapping will not happen here immediately, but will inevitably happen in the subsequent code flow

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:463]: GetServicesForPod retrieves the current llst of services associated with the pod. this only gives the list of added or updated resources. resources that previously had the mapping or those in the deleted case can only be retrieved from oldServicesList.

oldServicesList helps us map gateways that need to have the mappings removed

File: ako-gateway-api/k8s/gateway_controller.go
Comment (aaha97) [reply ako-gateway-api/k8s/gateway_controller.go:336]: this is simply the event handler for pod. updates to annotation will not trigger this event handler.

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (aaha97) [reply tests/gatewayapitests/npltests/gateway_npl_test.go]: changed to 60001

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (aaha97) [reply tests/gatewayapitests/npltests/gateway_npl_test.go]: changed to 60001

File: ako-gateway-api/k8s/gateway_controller.go
Comment (aaha97) [reply ako-gateway-api/k8s/gateway_controller.go]: fixed

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:439]: changed the log type to debug and fixed the wording

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Review by pkoshtavmware — DISMISSED** [review]
LGTM

File: internal/lib/lib.go
Comment (akshayhavile) [inline internal/lib/lib.go:1416]: This is not required as we have already appended endpointInformer. We should have if..elseif..else loop here. I think this happened after merge of endpointslice PR.

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (akshayhavile) [inline tests/gatewayapitests/npltests/gateway_npl_test.go:66]: We can move statement at Line 70 in place of 66. Use that variable in remaining places like line number 69, 71 etc.

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (akshayhavile) [inline tests/gatewayapitests/npltests/gateway_npl_test.go:78]: Should we disable DEBUG level?

**Review by akshayhavile — APPROVED** [review]
This comment can be taken in later PR. Approving now.

**Comment by akshayhavile** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (aaha97) [reply tests/gatewayapitests/npltests/gateway_npl_test.go:66]: line 66 and line 70 refer to different AKOControlConfig objects for ako and gateway-api respectively

File: tests/gatewayapitests/npltests/gateway_npl_test.go
Comment (aaha97) [reply tests/gatewayapitests/npltests/gateway_npl_test.go]: removed

File: ako-gateway-api/nodes/gateway_model_rel.go
Comment (aaha97) [reply ako-gateway-api/nodes/gateway_model_rel.go:473]: updated

File: internal/lib/lib.go
Comment (aaha97) [reply internal/lib/lib.go:1412]: fixed

**Comment by aaha97** [comment]
Adding latest UT results from local runs

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/k8stest       31.546s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/integrationtest       485.379s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/ingresstests  1014.492s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/oshiftroutetests      1314.548s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/bootuptests   21.363s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multicloudtests       171.543s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/servicesapitests      80.177s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/advl4tests    47.808s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/namespacesynctests    153.527s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/npltests      117.900s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests      1687.652s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests      1593.001s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedevhtests     693.597s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedevhtests     789.752s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedvstests      188.568s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/infratests    2.087s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/hatests       34.980s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/cnitests      104.664s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/cnitests      28.380s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/graphlayer    262.980s
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/ingestion     271.651s
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/npltests      63.692s
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/status        319.266s

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multitenancytests     71.948s

**Review by pkoshtavmware — APPROVED** [review]
LGTM

**Comment by aaha97** [comment]
build ako

**Review by akshayhavile — APPROVED** [review]

**Pull request merged**
